### ISSUE TITLE
Dropping empty map takes time proportional to capacity

### ISSUE THREAD

**Issue by jonhoo** [rust-lang/hashbrown#170]
We currently iterate over the map on drop:
https://github.com/rust-lang/hashbrown/blob/efbd03661fc9df0e594ecc89784107353cc71060/src/raw/mod.rs#L1194-L1195

And the iterator iterates over all the buckets, even if it knows there are no more items (`RawIter::items == 0`):
https://github.com/rust-lang/hashbrown/blob/efbd03661fc9df0e594ecc89784107353cc71060/src/raw/mod.rs#L1401-L1402

This means that code like this will unnecessarily take fairly long:
```rust
let map: HashMap<String, String> = HashMap::with_capacity(1024 * 1024);
drop(map);
```

The above is a little contrived, but the same occurs if the caller calls `drain` or `clear`, and then drops the map. Not sure if this is something we want to optimize for, but the fix is fairly straightforward.

**Comment by jonhoo** [comment]
I suppose this is mitigated by the fact that we iterate over the control bytes, not the actual buckets, so we're iterating through many buckets at once. Even so, it seems like it'd be nice to avoid this iteration altogether when we can?

**Comment by cuviper** [comment]
I guess the places that check `needs_drop` could easily check if it's already empty too.

**Comment by jonhoo** [comment]
Yup: https://github.com/jonhoo/hashbrown/commit/8dd041276bb58508f11de218b4187e393aff8e4c

**Comment by cuviper** [comment]
@jonhoo Are you going to submit that commit in a new pull request? I see that it was an incidental part of the closed PR #167, but that was not included in the replacement PR #175.

**Comment by jonhoo** [comment]
@cuviper Ah, good call! Submitted as #182.
